### PR TITLE
MGMT-4563 Skip host connection validation when host is installed or p…

### PR DESCRIPTION
**Pull request by oshercc**
At this point after rebooting the validation is still running so there is an option a user will get a disconnection or NTP validation fail post-installation, see [MGMT-4563](https://issues.redhat.com/browse/MGMT-4563) bug. 
This PR will skip connection validation checks post-reboot stage.

### CODE REVIEW

**Comment by ybettan** [comment]
I think that you can remove https://github.com/openshift/assisted-service/blob/913aba63aae8923400e8ff4e040c8cfc0d59897e/internal/host/host.go#L809-L811 as part of this PR.

If we ignore the `connected` validation after reboot, so it will never change, therefore, it will never be reported anyway - no need to recheck it in the failed-validations-report.

**Comment by oshercc** [comment]
> I think that you can remove
> 
> https://github.com/openshift/assisted-service/blob/913aba63aae8923400e8ff4e040c8cfc0d59897e/internal/host/host.go#L809-L811
> 
> as part of this PR.
> If we ignore the `connected` validation after reboot, so it will never change, therefore, it will never be reported anyway - no need to recheck it in the failed-validations-report.

@byount 
if this is removed the issue will not be resolved, see https://issues.redhat.com/browse/MGMT-4563

File: internal/host/refresh_status_preprocessor.go [hunk 1]
```diff
@@ -54,6 +55,12 @@ func (r *refreshPreprocessor) preprocess(c *validationContext) (map[string]bool,
 		st := v.condition(c)
 		stateMachineInput[v.id.String()] = st == ValidationSuccess
 		message := v.formatter(c, st)
+
+		// skip the fallowing validations when installation is done
+		if v.id == IsConnected && c.host.Progress.CurrentStage == models.HostStageDone {
```
Comment (tsorya) [inline internal/host/refresh_status_preprocessor.go:60]: is this enough? i think we should validate it when host got to reboot state. From this point there is no reason to validate if it is connected

File: internal/host/refresh_status_preprocessor.go [hunk 2]
```diff
@@ -54,6 +55,12 @@ func (r *refreshPreprocessor) preprocess(c *validationContext) (map[string]bool,
 		st := v.condition(c)
 		stateMachineInput[v.id.String()] = st == ValidationSuccess
 		message := v.formatter(c, st)
+
+		// skip the fallowing validations when installation is done
```
Comment (tsorya) [inline internal/host/refresh_status_preprocessor.go:59]: nit: typo following

**Comment by ybettan** [comment]
@oshercc 
look at the file name, this is not your code.
In the metrics/events report I already skip the `connect` validations after reboot and not reporting metrics/event in that case.

You change skip the validations computation itself in this case, therefore, the additional metrics/events check isn't necessary anymore an you can remove it.

**Comment by ybettan** [comment]
Also, can you please extend the commit message to explain why is this necessary ?

File: internal/host/refresh_status_preprocessor.go [hunk 3]
```diff
@@ -57,6 +63,12 @@ func (r *refreshPreprocessor) preprocess(c *validationContext) (map[string]bool,
 		st := v.condition(c)
 		stateMachineInput[v.id.String()] = st == ValidationSuccess
 		message := v.formatter(c, st)
+
+		// skip the fallowing validations when installation is post reboot
```
Comment (ybettan) [inline internal/host/refresh_status_preprocessor.go:67]: Don't we need to skip it before updating state-machine input as well?

File: subsystem/metrics_test.go [hunk 4]
```diff
@@ -314,27 +297,6 @@ var _ = Describe("Metrics tests", func() {
 			Expect(getValidationMetricCounter(string(models.HostValidationIDConnected), hostValidationFailedMetric)).To(Equal(oldFailedMetricCounter + 1))
 		})
 
-		It("'connected' failed after reboot", func() {
```
Comment (ybettan) [inline subsystem/metrics_test.go:317]: Why did you remove it? 
We can keep that test.

File: internal/host/refresh_status_preprocessor.go (hunk 3 above)
Comment (oshercc) [reply internal/host/refresh_status_preprocessor.go:67]: at this point, we only limit to those states and validations.

Comment (ybettan) [reply internal/host/refresh_status_preprocessor.go:67]: But why you skip only after updating `stateMachineInput[v.id.String()] = st == ValidationsSuccess` ?

Comment (oshercc) [reply internal/host/refresh_status_preprocessor.go:67]: stateMachineInput is used to compare with the previous status so in case there was a validation change there is a need to update

Comment (ybettan) [reply internal/host/refresh_status_preprocessor.go:67]: WTYT @filanov is this a problem to skip the validation in the validation_output but not in the statemachine_input?
This will probably create some inconsistencies no?!

Comment (filanov) [reply internal/host/refresh_status_preprocessor.go:67]: What will happen if you skip `stateMachineInput` as well? 
lets say that once you skip this validation in the output then the new validation will not contain the skipped validations so `didValidationChanged` will detect a diff, it will update the DB but will not send any event `reportValidationStatusChanged` iterate over the new conditions.
So why not to skip the state machine input?

Comment (ybettan) [reply internal/host/refresh_status_preprocessor.go:67]: > So why not to skip the state machine input?

This is what I have suggested, I asked if to skip it in the state machine input can create another issue that I am not aware of.

Anyway we got our answer.
@oshercc I think we need to skip it in the state machine input as well.

Comment (filanov) [reply internal/host/refresh_status_preprocessor.go:67]: There may be a problem if state machine use those conditions, So maybe instead of skipping we should mark them as pass or not relevant?

Comment (filanov) [reply internal/host/refresh_status_preprocessor.go:67]: Actually marking them as passed will be misleading.

Comment (oshercc) [reply internal/host/refresh_status_preprocessor.go:67]: how can I mark them as `not relevant ? that will require a third state that is not boolean

Comment (oshercc) [reply internal/host/refresh_status_preprocessor.go:67]: @filanov maybe I didn't understand what you mean by `not relevant do you mean an additional state or is there a way of doing it?

Comment (ybettan) [reply internal/host/refresh_status_preprocessor.go:67]: I think he was referring to `ValidationSuccess`, `ValidationFailure`, `ValidationPending` and add a 4th option but those are for the validationsOutput and not for the stateMachineInput.

Comment (filanov) [reply internal/host/refresh_status_preprocessor.go:67]: It can be an option. @avishayt @ori-amizur what do you think?

Comment (ori-amizur) [reply internal/host/refresh_status_preprocessor.go:67]: Maybe we want to mark in the validation what are the relevant stages/statuses for the validation?

Comment (ori-amizur) [reply internal/host/refresh_status_preprocessor.go:67]: We can a field in validation https://github.com/openshift/assisted-service/blob/69594371fa01cd67db43093542ee5f6fbd449134/internal/host/validator.go#L62
that may be a function, that indicates if the validation should be evaluated

**Comment by jordigilh** [comment]
Any idea if this is going to impact the UI? I'm working on another [PR](https://github.com/openshift/assisted-service/pull/1554) to "disable" validations (not exactly this one, but similar), and my first approach was to remove the validations from the state machine and the slice of validations. It turned out that the UI expects these validations to be included in the results of the REST call. The proposed new approach was then to create a new status "disabled" that the UI will parse and act accordingly. Would this idea work here as well? /cc @mareklibra @jtomasek
